linkerd/retry/src/replay.rs

```diff
@@ -192,10 +208,23 @@ impl<B: Body + Unpin> Body for ReplayBody<B> {
     }
 
     fn is_end_stream(&self) -> bool {
-        self.state
-            .as_ref()
-            .and_then(|state| state.rest.as_ref().map(Body::is_end_stream))
-            .unwrap_or(false)
+        // if the initial body was EOS as soon as it was wrapped, then we are
+        // empty.
+        if self.shared.was_empty {
+            return true;
+        }
+
+        // if this body has data or trailers remaining to play back, it
+        // is not EOS
+        !self.replay_body
+            && !self.replay_trailers
+            // or, if we have replayed everything, the initial body may
+            // still have data remaining, so ask it
+            && self
```

Comment:
the comment says "or" but the condition is `&&`... won't this always check the initial body? is that necessary/correct?

if so, perhaps extract the lower inner-body check into a temp variable so the condition reads more simply as

```rust
let inner_is_end = ...;

!self.replay_body && !self.replay_trailers && inner_is_end
```

Reply:
> won't this always check the initial body? is that necessary/correct?

yes, we may have replayed all of the buffered data from a body that still has data remaining, and we can't know for certain unless we have checked the initial body

> if so, perhaps extract the lower inner-body check into a temp variable so the condition reads more simply as

sure, can do, i didn't do that because i thought it was better to allow it to short-circuit and avoid checking the inner body at all if we haven't replayed any data yet? this probably isn't hot enough that it's important for performance reasons, but i thought it was slightly better that way...